[WIP] Require transactions by default for all Hibernate ORM session use

yrodiere opened the pull request:
* Fixes #51244

Please do not merge, I expect many tests will fail but I want to get a sense of how many.

We will need an entry in the migration guide.

cc @Sanne @gavinking 

See also https://hibernate.zulipchat.com/#narrow/channel/132094-hibernate-orm-dev/topic/StatelessSession.20insertMultiple.20.2F.20updateMultiple.20.2F.20etc.2E/near/538862234

By the way, here is the relevant code:

https://github.com/quarkusio/quarkus/blob/e16597b7d4345e456df822cd0df6255fb9eb5a3e/extensions/hibernate-orm/runtime/src/main/java/io/quarkus/hibernate/orm/runtime/session/TransactionScopedSession.java#L116-L133

yrodiere committed:
Require transactions by default for all Hibernate ORM session use

Assisted-By: Claude Sonnet 4.5 <noreply@anthropic.com>

Sanne commented:
> BTW, shouldn't we also require transactions for HR, then? Just to be consistent.

Perhaps yes, but bear in mind in that case the mechanism is quite different as it wouldn't be JTA. We'd need to think of the consequences of that - I'd be especially more concerned on starting them automatically (w/o an explicit opt-in) as we'd not want to give a false sense of safety w/o the user understanding the implications.

yrodiere commented:
> This is becoming a 2-factor API: you can only use it with an annotation and using its API.

Technically you can also start a transaction programmatically with `QuarkusTransaction`. But yes, essentially, it is 2-factor: you need to specify the scope of whatever session you're using, and that requires specifying the transaction boundaries.

> I do think we are going to have to talk about simplifying access to the ORM API, even though this can be done later, but perhaps REST endpoints by default should have an automatic `@Transactional`, assuming this works for both ORM and HR. In the future.

That would cause a lot of backwards incompatibility since it would effectively extend the scope of `@Transactional` methods called by REST endpoints.

We could instead have some sort of "last-minute" default transaction scoping that, assuming a request scope exists and no transaction scope exists when first accessing the `Session`, would start a transaction scope automatically and make sure it ends (commit/rollback) when the request scope ends.

But:

1. It would work for Hibernate, but not for direct use of JDBC or Reactive SQL clients -- since we allow no-transaction access there.
2. It feels very, very magic at this point. So much so that I shudder to think of the experience of a user trying to debug problems related to this behavior.
3. Alternatively if you want to keep things simple you can just put `@Transactional` on your REST resources, at the class level. And then you're all set. It's not that complex.

yrodiere commented:
> @yrodiere What's the status on this one?

Many tests failing, as you can see.

@Sanne suggested we'd better be damn sure use all use cases work fine with transactions before we do this, which I agree with. There are still edge cases where not using a transaction is... arguably correct.

So, we'll need to address a good chunk of https://github.com/quarkusio/quarkus/issues/51243 before we flip the switch here.

yrodiere committed:
Refactor Hibernate ORM CDI tests to support both request-scoped scenarios

Split CDI-based persistence unit tests into RequestScopeEnabled and RequestScopeDisabled
variants to explicitly test both behaviors:
- RequestScopeEnabled: reads work without transaction, writes require transaction
- RequestScopeDisabled: both reads and writes require transaction

This refactoring prepares for the upcoming change that will make request-scoped disabled
by default. The RequestScopeEnabled variants currently assume the default (enabled), while
RequestScopeDisabled variants explicitly disable it via config.

Deleted original test classes (now replaced by RequestScopeEnabled variants):
- MultiplePersistenceUnitsCdiEntityManagerTest
- MultiplePersistenceUnitsCdiSessionTest
- SinglePersistenceUnitCdiEntityManagerTest
- SinglePersistenceUnitCdiSessionTest

Deleted redundant tests (covered by new variants):
- SessionWithinRequestScopeTest/Disabled (uses same @ActivateRequestContext pattern)
- StatelessSessionWithinRequestScopeTest/Disabled (same pattern)
- NoTransactionTest (covered by NoRequestNoTransaction tests)

Assisted-By: Claude Sonnet 4.5 <noreply@anthropic.com>

yrodiere committed:
Fix Hibernate ORM tests not using transactions for no reason

Assisted-By: Claude Sonnet 4.5 <noreply@anthropic.com>
